File: packages/license/src/token.ts
export const TOKEN_PREFIX = "tnlic1";
export const LICENSE_TYP = "tunnet-license+2";
export const REVOCATION_TYP = "tunnet-revocations+2";

export const MAX_TOKEN_CHARS = 8192;
const MAX_SEGMENT_CHARS = 4096;
const B64URL = /^[A-Za-z0-9_-]+$/;

const enc = new TextEncoder();
const dec = new TextDecoder("utf-8", { fatal: true });

export function utf8(s: string): Uint8Array {
  return enc.encode(s);
}

export function b64uEncode(bytes: Uint8Array): string {
  let bin = "";
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]!);
  return btoa(bin).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

export function b64uDecode(value: string): Uint8Array {
  if (!B64URL.test(value) || value.length % 4 === 1) {
    throw new SyntaxError("invalid base64url");
  }
  const padded =
    value.replace(/-/g, "+").replace(/_/g, "/") +
    "=".repeat((4 - (value.length % 4)) % 4);
  const bin = atob(padded);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  if (b64uEncode(out) !== value)
    throw new SyntaxError("non-canonical base64url");
  return out;
}

export type ProtectedHeader = {
  readonly alg: "Ed25519";
  readonly kid: string;
  readonly typ: string;
  readonly crit: readonly string[];
};

export type DecodedToken = {
  readonly header: ProtectedHeader;
  readonly payload: Record<string, unknown>;
  readonly signature: Uint8Array;
  readonly signingInput: Uint8Array;
};

function parseJsonObject(bytes: Uint8Array): Record<string, unknown> {
  const value: unknown = JSON.parse(dec.decode(bytes));
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new SyntaxError("expected JSON object");
  }
  return value as Record<string, unknown>;
}

function parseHeader(raw: Record<string, unknown>): ProtectedHeader {
  if (raw.alg !== "Ed25519") throw new SyntaxError("unsupported alg");
  if (
    typeof raw.kid !== "string" ||
    raw.kid.length === 0 ||
    raw.kid.length > 64
  ) {
    throw new SyntaxError("invalid kid");
  }
  if (typeof raw.typ !== "string") throw new SyntaxError("invalid typ");
  const crit = raw.crit === undefined ? [] : raw.crit;
  if (!Array.isArray(crit) || crit.some((c) => typeof c !== "string")) {
    throw new SyntaxError("invalid crit");
  }
  return Object.freeze({
    alg: "Ed25519",
    kid: raw.kid,
    typ: raw.typ,
    crit: Object.freeze([...crit] as string[]),
  });
}

export function decodeToken(token: string): DecodedToken {
  const trimmed = token.trim();
  if (trimmed.length > MAX_TOKEN_CHARS) throw new RangeError("token too large");

  const parts = trimmed.split(".");
  if (parts.length !== 4 || parts[0] !== TOKEN_PREFIX) {
    throw new SyntaxError("unsupported token format");
  }
  const [, h, p, s] = parts as [string, string, string, string];
  if (
    h.length > MAX_SEGMENT_CHARS ||
    p.length > MAX_SEGMENT_CHARS ||
    s.length > 128
  ) {
    throw new RangeError("segment too large");
  }

  const signature = b64uDecode(s);
  if (signature.length !== 64)
    throw new SyntaxError("invalid signature length");

  return Object.freeze({
    header: parseHeader(parseJsonObject(b64uDecode(h))),
    payload: parseJsonObject(b64uDecode(p)),
    signature,
    signingInput: utf8(`${TOKEN_PREFIX}.${h}.${p}`),
  });
}

export function buildSigningInput(
  header: object,
  payload: object,
): { input: Uint8Array; h: string; p: string } {
  const h = b64uEncode(utf8(JSON.stringify(header)));
  const p = b64uEncode(utf8(JSON.stringify(payload)));
  return { input: utf8(`${TOKEN_PREFIX}.${h}.${p}`), h, p };
}

export function assembleToken(
  h: string,
  p: string,
  signature: Uint8Array,
): string {
  return `${TOKEN_PREFIX}.${h}.${p}.${b64uEncode(signature)}`;
}

File: crates/tunnet-license/src/token.rs
//! tnlic1 token decode (header / payload / signature).

#[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
use std::sync::LazyLock;

use base64::Engine;
#[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
use base64::engine::general_purpose::NO_PAD;
#[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
use base64::engine::general_purpose::URL_SAFE_NO_PAD as B64U_ENGINE;
#[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
use base64::engine::simd::Simd;
use serde::Deserialize;
use serde_json::Value;

use crate::error::LicenseError;

#[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
static B64U: LazyLock<Simd> = LazyLock::new(|| Simd::url_safe(NO_PAD));

pub const TOKEN_PREFIX: &str = "tnlic1";
pub const LICENSE_TYP: &str = "tunnet-license+2";
pub const MAX_TOKEN_LEN: usize = 8192;
const MAX_SEGMENT_CHARS: usize = 4096;

#[derive(Debug, Clone)]
pub struct ProtectedHeader {
    pub alg: String,
    pub kid: String,
    pub typ: String,
    pub crit: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct DecodedToken {
    pub header: ProtectedHeader,
    pub payload: Value,
    pub signature: [u8; 64],
    /// Original `tnlic1.{h}.{p}` bytes (not re-serialized).
    pub signing_input: Vec<u8>,
}

fn is_b64url_char(c: u8) -> bool {
    c.is_ascii_alphanumeric() || c == b'-' || c == b'_'
}

fn b64u_engine_encode(bytes: &[u8]) -> String {
    #[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
    {
        B64U.encode(bytes)
    }
    #[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
    {
        B64U_ENGINE.encode(bytes)
    }
}

fn b64u_engine_decode(value: &str) -> Result<Vec<u8>, base64::DecodeError> {
    #[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
    {
        B64U.decode(value)
    }
    #[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
    {
        B64U_ENGINE.decode(value)
    }
}

/// Strict canonical base64url decode (no padding; round-trip must match).
pub fn b64u_decode(value: &str) -> Result<Vec<u8>, LicenseError> {
    if value.as_bytes().iter().any(|b| !is_b64url_char(*b)) || value.len() % 4 == 1 {
        return Err(LicenseError::malformed("invalid base64url"));
    }
    let bytes =
        b64u_engine_decode(value).map_err(|_| LicenseError::malformed("invalid base64url"))?;
    let re = b64u_engine_encode(&bytes);
    if re != value {
        return Err(LicenseError::malformed("non-canonical base64url"));
    }
    Ok(bytes)
}

pub fn b64u_encode(bytes: &[u8]) -> String {
    b64u_engine_encode(bytes)
}

fn parse_json_object(bytes: &[u8]) -> Result<Value, LicenseError> {
    let value: Value =
        serde_json::from_slice(bytes).map_err(|e| LicenseError::malformed(format!("json: {e}")))?;
    if !value.is_object() {
        return Err(LicenseError::malformed("expected JSON object"));
    }
    Ok(value)
}

#[derive(Debug, Deserialize)]
struct RawHeader {
    alg: String,
    kid: String,
    typ: String,
    #[serde(default)]
    crit: Vec<String>,
}

fn parse_header(raw: Value) -> Result<ProtectedHeader, LicenseError> {
    let h: RawHeader =
        serde_json::from_value(raw).map_err(|e| LicenseError::malformed(format!("header: {e}")))?;
    if h.alg != "Ed25519" {
        return Err(LicenseError::malformed("unsupported alg"));
    }
    if h.kid.is_empty() || h.kid.len() > 64 {
        return Err(LicenseError::malformed("invalid kid"));
    }
    Ok(ProtectedHeader {
        alg: h.alg,
        kid: h.kid,
        typ: h.typ,
        crit: h.crit,
    })
}

pub fn decode_token(token: &str) -> Result<DecodedToken, LicenseError> {
    let trimmed = token.trim();
    if trimmed.len() > MAX_TOKEN_LEN {
        return Err(LicenseError::too_large("token too large"));
    }

    let parts: Vec<&str> = trimmed.split('.').collect();
    if parts.len() != 4 || parts[0] != TOKEN_PREFIX {
        return Err(LicenseError::malformed("unsupported token format"));
    }
    let (h, p, s) = (parts[1], parts[2], parts[3]);
    if h.len() > MAX_SEGMENT_CHARS || p.len() > MAX_SEGMENT_CHARS || s.len() > 128 {
        return Err(LicenseError::too_large("segment too large"));
    }

    let sig_bytes = b64u_decode(s)?;
    let signature: [u8; 64] = sig_bytes
        .as_slice()
        .try_into()
        .map_err(|_| LicenseError::malformed("invalid signature length"))?;

    let header = parse_header(parse_json_object(&b64u_decode(h)?)?)?;
    let payload = parse_json_object(&b64u_decode(p)?)?;
    let signing_input = format!("{TOKEN_PREFIX}.{h}.{p}").into_bytes();

    Ok(DecodedToken {
        header,
        payload,
        signature,
        signing_input,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    #[case::empty("")]
    #[case::not_a_token("not-a-token")]
    #[case::wrong_prefix("tnlic0.a.b.c")]
    #[case::missing_signature("tnlic1.a.b")]
    #[case::extra_segment("tnlic1.a.b.c.d")]
    #[case::padding_is_forbidden("tnlic1.YQ==.e30.AA")]
    fn rejects_malformed_token_envelopes(#[case] token: &str) {
        assert!(decode_token(token).is_err());
    }

    #[test]
    fn rejects_too_large() {
        let huge = format!("tnlic1.{}.b.c", "a".repeat(MAX_TOKEN_LEN));
        let err = decode_token(&huge).unwrap_err();
        assert_eq!(err.code, crate::error::LicenseFailureCode::TooLarge);
    }
}
